File: web/src/domain/attendance.ts
import type { AttendanceStatus, NewAttendanceRecord, Schedule } from "./model";
import { recordKey } from "./model";
import type { Id, IsoDate } from "./primitives";
import { timeToMinutes } from "./time";

/**
 * The daily lifecycle of one schedule.
 *
 *   before ──▶ present ──▶ [gap] ──▶ late ──▶ closed
 *
 * `gap` exists only when a schedule leaves a deliberate pause between the two windows; when the
 * late window starts exactly as the present window ends, the state machine never enters it.
 * A scan is graded by whichever window is open. Scans in `before`, `gap`, or `closed` record
 * nothing at all — matching the original app, where a scan outside both windows was discarded.
 */
export type ScheduleWindowState = "before" | "present" | "gap" | "late" | "closed";

export function windowStateAt(schedule: Schedule, atMinutes: number): ScheduleWindowState {
  const presentStart = timeToMinutes(schedule.present.start);
  const presentEnd = timeToMinutes(schedule.present.end);
  const lateStart = timeToMinutes(schedule.late.start);
  const lateEnd = timeToMinutes(schedule.late.end);

  if (atMinutes < presentStart) return "before";
  if (atMinutes < presentEnd) return "present";
  if (atMinutes < lateStart) return "gap";
  if (atMinutes < lateEnd) return "late";
  return "closed";
}

/** The status a scan earns in a given state, or null when the scan records nothing. */
export function statusForWindow(state: ScheduleWindowState): AttendanceStatus | null {
  if (state === "present") return "present";
  if (state === "late") return "late";
  return null;
}

/**
 * Minute at which this schedule next changes state, or null once it is closed for the day.
 * The UI uses this to wake up exactly when the late window shuts and run the absentee sweep,
 * instead of polling.
 */
export function nextTransitionMinute(schedule: Schedule, atMinutes: number): number | null {
  const boundaries = [
    timeToMinutes(schedule.present.start),
    timeToMinutes(schedule.present.end),
    timeToMinutes(schedule.late.start),
    timeToMinutes(schedule.late.end),
  ].sort((a, b) => a - b);

  for (const boundary of boundaries) {
    if (boundary > atMinutes) return boundary;
  }
  return null;
}

export function isCollectingScans(schedule: Schedule, atMinutes: number): boolean {
  return statusForWindow(windowStateAt(schedule, atMinutes)) !== null;
}

/* --------------------------------------------------------------- Scan intake */

export interface ScanContext {
  studentNumber: string;
  /** Active schedules belonging to the scanned student's section. */
  schedules: readonly Schedule[];
  /** Keys of records that already exist, from {@link recordKey}. */
  existingKeys: ReadonlySet<string>;
  date: IsoDate;
  atMinutes: number;
  recordedAt: string;
}

export interface ScanResolution {
  /** Records to append. Empty when nothing about this scan was new. */
  created: NewAttendanceRecord[];
  /** Schedules the student was already recorded for today — a second scan changes nothing. */
  duplicateScheduleIds: Id[];
  /** Schedules whose windows were shut at scan time. */
  inactiveScheduleIds: Id[];
}

/**
 * Grades one scan against every active schedule for the student's section.
 *
 * A student may sit in more than one schedule at once (a class and a flag ceremony, say), so a
 * single scan can produce several records. Duplicate suppression is per schedule, not per scan:
 * scanning again after a new window opens correctly records the new window while leaving the
 * already-recorded one untouched.
 */
export function resolveScan(context: ScanContext): ScanResolution {
  const resolution: ScanResolution = {
    created: [],
    duplicateScheduleIds: [],
    inactiveScheduleIds: [],
  };

  for (const schedule of context.schedules) {
    if (schedule.id === undefined || schedule.archived) continue;

    const status = statusForWindow(windowStateAt(schedule, context.atMinutes));
    if (status === null) {
      resolution.inactiveScheduleIds.push(schedule.id);
      continue;
    }

    const key = recordKey({
      studentNumber: context.studentNumber,
      scheduleId: schedule.id,
      date: context.date,
    });
    if (context.existingKeys.has(key)) {
      resolution.duplicateScheduleIds.push(schedule.id);
      continue;
    }

    resolution.created.push({
      scheduleId: schedule.id,
      sectionId: schedule.sectionId,
      studentNumber: context.studentNumber,
      date: context.date,
      status,
      scheduleTitle: schedule.title,
      recordedAt: context.recordedAt,
    });
  }

  return resolution;
}

/* ------------------------------------------------------------ Absentee sweep */

export interface SweepContext {
  schedule: Schedule;
  /** Student numbers of every active student in the schedule's section. */
  activeStudentNumbers: readonly string[];
  existingKeys: ReadonlySet<string>;
  date: IsoDate;
  atMinutes: number;
  recordedAt: string;
}

/**
 * Once the late window has shut, every active student in the section with no record for this
 * schedule and date is absent.
 *
 * Idempotent by construction: the records it writes are exactly the ones whose absence it
 * detects, so a second run finds nothing left to write. Returns [] while the schedule is still
 * open — sweeping early would mark students absent who have not had their chance to scan.
 */
export function absentRecordsFor(context: SweepContext): NewAttendanceRecord[] {
  const { schedule } = context;
  if (schedule.id === undefined || schedule.archived) return [];
  if (windowStateAt(schedule, context.atMinutes) !== "closed") return [];

  const scheduleId = schedule.id;

  return context.activeStudentNumbers
    .filter(
      (studentNumber) =>
        !context.existingKeys.has(recordKey({ studentNumber, scheduleId, date: context.date })),
    )
    .map((studentNumber) => ({
      scheduleId,
      sectionId: schedule.sectionId,
      studentNumber,
      date: context.date,
      status: "absent" as const,
      scheduleTitle: schedule.title,
      recordedAt: context.recordedAt,
    }));
}

File: mobile-android/app/src/main/java/ph/attendance/domain/Attendance.kt
package ph.attendance.domain

/**
 * The daily lifecycle of one schedule.
 *
 *   BEFORE ──▶ PRESENT ──▶ [GAP] ──▶ LATE ──▶ CLOSED
 *
 * `GAP` exists only when a schedule leaves a deliberate pause between the two windows; when the
 * late window opens exactly as the present window closes, the machine never enters it. A scan is
 * graded by whichever window is open, and scans in BEFORE, GAP, or CLOSED record nothing at all —
 * matching the original app, where a scan outside both windows was discarded.
 */
enum class ScheduleWindowState { BEFORE, PRESENT, GAP, LATE, CLOSED }

fun windowStateAt(schedule: Schedule, atMinutes: Int): ScheduleWindowState {
    val presentStart = timeToMinutes(schedule.present.start)
    val presentEnd = timeToMinutes(schedule.present.end)
    val lateStart = timeToMinutes(schedule.late.start)
    val lateEnd = timeToMinutes(schedule.late.end)

    return when {
        atMinutes < presentStart -> ScheduleWindowState.BEFORE
        atMinutes < presentEnd -> ScheduleWindowState.PRESENT
        atMinutes < lateStart -> ScheduleWindowState.GAP
        atMinutes < lateEnd -> ScheduleWindowState.LATE
        else -> ScheduleWindowState.CLOSED
    }
}

/** The status a scan earns in a given state, or null when the scan records nothing. */
fun statusForWindow(state: ScheduleWindowState): AttendanceStatus? = when (state) {
    ScheduleWindowState.PRESENT -> AttendanceStatus.PRESENT
    ScheduleWindowState.LATE -> AttendanceStatus.LATE
    else -> null
}

fun isCollectingScans(schedule: Schedule, atMinutes: Int): Boolean =
    statusForWindow(windowStateAt(schedule, atMinutes)) != null

/**
 * The minute this schedule next changes state, or null once it is closed for the day. The UI uses
 * it to refresh exactly when a window opens or shuts instead of polling.
 */
fun nextTransitionMinute(schedule: Schedule, atMinutes: Int): Int? = listOf(
    timeToMinutes(schedule.present.start),
    timeToMinutes(schedule.present.end),
    timeToMinutes(schedule.late.start),
    timeToMinutes(schedule.late.end),
).sorted().firstOrNull { it > atMinutes }

data class ScanContext(
    val studentNumber: String,
    /** Active schedules belonging to the scanned student's section. */
    val schedules: List<Schedule>,
    val existingKeys: Set<String>,
    val date: String,
    val atMinutes: Int,
    val recordedAt: String,
)

data class ScanResolution(
    /** Records to append. Empty when nothing about this scan was new. */
    val created: List<AttendanceRecord>,
    /** Schedules already recorded today — a second scan changes nothing. */
    val duplicateScheduleIds: List<Id>,
    /** Schedules whose windows were shut at scan time. */
    val inactiveScheduleIds: List<Id>,
)

/**
 * Grades one scan against every active schedule for the student's section.
 *
 * A student can sit in more than one schedule at once — a class and a flag ceremony — so a single
 * scan may produce several records. Duplicate suppression is per schedule rather than per scan,
 * so scanning again once a new window opens records the new window and leaves the old one alone.
 */
fun resolveScan(context: ScanContext): ScanResolution {
    val created = mutableListOf<AttendanceRecord>()
    val duplicates = mutableListOf<Id>()
    val inactive = mutableListOf<Id>()

    for (schedule in context.schedules) {
        if (schedule.archived) continue

        val status = statusForWindow(windowStateAt(schedule, context.atMinutes))
        if (status == null) {
            inactive += schedule.id
            continue
        }

        if (recordKey(context.studentNumber, schedule.id, context.date) in context.existingKeys) {
            duplicates += schedule.id
            continue
        }

        created += AttendanceRecord(
            id = newId(),
            scheduleId = schedule.id,
            sectionId = schedule.sectionId,
            studentNumber = context.studentNumber,
            date = context.date,
            status = status,
            scheduleTitle = schedule.title,
            recordedAt = context.recordedAt,
        )
    }

    return ScanResolution(created, duplicates, inactive)
}

data class SweepContext(
    val schedule: Schedule,
    /** Student numbers of every active student in the schedule's section. */
    val activeStudentNumbers: List<String>,
    val existingKeys: Set<String>,
    val date: String,
    val atMinutes: Int,
    val recordedAt: String,
)

/**
 * Once the late window has shut, every active student in the section with no record for this
 * schedule and date is absent.
 *
 * Idempotent by construction: it writes exactly the records whose absence it detects, so a second
 * run finds nothing left to write. Returns nothing while the schedule is still open — sweeping
 * early would mark students absent who have not yet had their chance to scan.
 */
fun absentRecordsFor(context: SweepContext): List<AttendanceRecord> {
    val schedule = context.schedule
    if (schedule.archived) return emptyList()
    if (windowStateAt(schedule, context.atMinutes) != ScheduleWindowState.CLOSED) return emptyList()

    return context.activeStudentNumbers
        .filter { recordKey(it, schedule.id, context.date) !in context.existingKeys }
        .map { studentNumber ->
            AttendanceRecord(
                id = newId(),
                scheduleId = schedule.id,
                sectionId = schedule.sectionId,
                studentNumber = studentNumber,
                date = context.date,
                status = AttendanceStatus.ABSENT,
                scheduleTitle = schedule.title,
                recordedAt = context.recordedAt,
            )
        }
}
